packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol

```diff
@@ -426,70 +433,66 @@ contract PerpetualLiquidatable is PerpetualPositionManager {
-            ) {
-                delete liquidations[sponsor][liquidationId];
-            }
+            // If the dispute is successful then all three users should receive rewards:
+
+            // Pay DISPUTER: disputer reward + dispute bond + returned final fee
+            rewards.payToDisputer = disputerDisputeReward.add(disputeBondAmount).add(finalFee);
+
+            // Pay SPONSOR: remaining collateral (collateral - TRV) + sponsor reward
+            rewards.payToSponsor = sponsorDisputeReward.add(collateral.sub(tokenRedemptionValue));
+
+            // Pay LIQUIDATOR: TRV - dispute reward - sponsor reward
+            // If TRV > Collateral, then subtract rewards from collateral
+            // NOTE: This should never be below zero since we prevent (sponsorDisputePct+disputerDisputePct) >= 0 in
+            // the constructor when these params are set.
+            rewards.payToLiquidator = tokenRedemptionValue.sub(sponsorDisputeReward).sub(disputerDisputeReward);
+
             // In the case of a failed dispute only the liquidator can withdraw.
-        } else if (liquidation.state == Status.DisputeFailed && msg.sender == liquidation.liquidator) {
+        } else if (liquidation.state == Status.DisputeFailed) {
             // Pay LIQUIDATOR: collateral + dispute bond + returned final fee
-            withdrawalAmount = collateral.add(disputeBondAmount).add(finalFee);
-            delete liquidations[sponsor][liquidationId];
+            rewards.payToLiquidator = collateral.add(disputeBondAmount).add(finalFee);
+
             // If the state is pre-dispute but time has passed liveness then there was no dispute. We represent this
             // state as a dispute failed and the liquidator can withdraw.
-        } else if (liquidation.state == Status.PreDispute && msg.sender == liquidation.liquidator) {
+        } else if (liquidation.state == Status.PreDispute) {
             // Pay LIQUIDATOR: collateral + returned final fee
-            withdrawalAmount = collateral.add(finalFee);
-            delete liquidations[sponsor][liquidationId];
+            rewards.payToLiquidator = collateral.add(finalFee);
         }
-        require(withdrawalAmount.isGreaterThan(0), "Invalid withdrawal amount");
 
-        // Decrease the total collateral held in liquidatable by the amount withdrawn.
-        amountWithdrawn = _removeCollateral(rawLiquidationCollateral, withdrawalAmount);
+        require(
+            (rewards.payToLiquidator.add(rewards.payToSponsor).add(rewards.payToDisputer)).isGreaterThan(0),
+            "Nothing to withdraw"
+        );
 
-        emit LiquidationWithdrawn(msg.sender, amountWithdrawn.rawValue, liquidation.state, settlementPrice.rawValue);
+        // Decrease the total collateral held in liquidatable by the amount to pay each party. The actual amounts
+        // withdrawn might differ if _removeCollateral causes precision loss.
+        if (rewards.payToLiquidator.isGreaterThan(0)) {
```

Comment:
Seeing there are only three branches that payment can occur above (`DisputeSucceeded`, `DisputeFailed ` or `PreDispute`) would it make sense to remove these if statements and simply put the logic to decrement the total collateral and preform the token transfers within the respective if/else branches above? If you did this you'd be able to remove 3 if statements. WDYT?

Reply:
You're totally right I'll try that

Reply:
Addressed in this [commit](https://github.com/UMAprotocol/protocol/pull/2152/commits/ba30b32ddad538de6e07dadfff09e7ace7e8989f)